src/services/classifier2020.ts

```diff
@@ -0,0 +1,257 @@
+/** @internal */
+namespace ts.classifier.v2020 {
+
+    /** @internal */
+    export const enum TokenEncodingConsts {
+        typeOffset = 8,
+        modifierMask = (1 << typeOffset) - 1
+    }
+
+    /** @internal */
+    export const enum TokenType {
+        class, enum, interface, namespace, typeParameter, type, parameter, variable, enumMember, property, function, member
+    }
+
+    /** @internal */
+    export const enum TokenModifier {
+        declaration, static, async, readonly, defaultLibrary, local
+    }
+
+    /** This is mainly used internally for testing */
+    export function getSemanticClassifications(program: Program, _cancellationToken: CancellationToken, sourceFile: SourceFile, span: TextSpan): ClassifiedSpan[] {
+        const classifications = getEncodedSemanticClassifications(program, _cancellationToken, sourceFile, span);
+
+        Debug.assert(classifications.spans.length % 3 === 0);
+        const dense = classifications.spans;
+        const result: ClassifiedSpan[] = [];
+        for (let i = 0; i < dense.length; i += 3) {
+            result.push({
+                textSpan: createTextSpan(dense[i], dense[i + 1]),
+                classificationType: dense[i + 2]
+            });
+        }
+
+        return result;
+    }
+
+    export function getEncodedSemanticClassifications(program: Program, _cancellationToken: CancellationToken, sourceFile: SourceFile, span: TextSpan): Classifications {
+        return {
+            spans: getSemanticTokens(program, sourceFile, span),
+            endOfLineState: EndOfLineState.None
+        };
+    }
+
+    function getSemanticTokens(program: Program, sourceFile: SourceFile, span: TextSpan): number[] {
+        const resultTokens: number[] = [];
+
+        const collector = (node: Node, typeIdx: number, modifierSet: number) => {
+            resultTokens.push(node.getStart(sourceFile), node.getWidth(sourceFile), ((typeIdx + 1) << TokenEncodingConsts.typeOffset) + modifierSet);
+        };
+
+        if (program && sourceFile) {
+            collectTokens(program, sourceFile, span, collector);
+        }
+        return resultTokens;
+    }
+
+    function collectTokens(program: Program, sourceFile: SourceFile, span: TextSpan, collector: (node: Node, tokenType: number, tokenModifier: number) => void) {
+        const typeChecker = program.getTypeChecker();
+
+        let inJSXElement = false;
+
+        function visit(node: Node) {
+            if (!node || !textSpanIntersectsWith(span, node.pos, node.getFullWidth()) || node.getFullWidth() === 0) {
+                return;
+            }
+            const prevInJSXElement = inJSXElement;

```

Comment:
You need to check for cancellation token here. Generally we do that if for 
```
                    case SyntaxKind.ModuleDeclaration:
                    case SyntaxKind.ClassDeclaration:
                    case SyntaxKind.InterfaceDeclaration:
                    case SyntaxKind.FunctionDeclaration:
                    case SyntaxKind.ClassExpression:
                    case SyntaxKind.FunctionExpression:
                    case SyntaxKind.ArrowFunction:
```

Reply:
This is in 👍